misc/python/materialize/cli/orchestratord.py

```diff
@@ -215,74 +216,124 @@ def root_psql(cmd: str):
+            f"deployment/mz{resource_id}-console",
+        ),
+        exception_types=(subprocess.CalledProcessError,),
+    )
 
-        if resource_id is not None:
-            break
 
-        sleep(1)
-    assert resource_id is not None
+def portforward(args: argparse.Namespace):
+    env_kubectl = make_env_kubectl(args)
 
-    node_port = None
-    for _ in range(60):
-        try:
-            node_port = (
-                env_kubectl(
-                    "get",
-                    "service",
-                    f"mz{resource_id}-console",
-                    '-o=jsonpath={.spec.ports[?(@.name=="http")].nodePort}',
-                )
-                .decode()
-                .strip()
+    resource_id = get_resource_id(args)
+
+    sockets = []
+    try:
+        for _ in range(5):
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sockets.append(s)
+            s.bind(("", 0))
+            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        ports = [s.getsockname()[1] for s in sockets]
+    finally:
+        for s in sockets:
+            s.close()
+
+    port_destinations = {}
+
+    def port_forward(target: str, target_port: str):
+        port = ports.pop()
+        port_destinations[f"{target}/{target_port}"] = port
+
+        # in theory doing a port forward to the service directly should work,
+        # but i can't seem to get it to
+        selector = json.loads(
+            env_kubectl(
+                "get",
+                "service",
+                f"mz{resource_id}-{target}",
+                "-o=jsonpath={.spec.selector}",
             )
-            if node_port == "<no value>" or node_port == "":
-                node_port = None
-        except subprocess.CalledProcessError:
-            node_port = None
+        )
+        pod = env_kubectl(
+            "get",
+            "pod",
+            "-o=name",
+            *[f"--selector={k}={v}" for k, v in selector.items()],
+        ).splitlines()[0]
 
-        if node_port is not None:
-            break
+        env_kubectl(
+            "port-forward",
+            pod,
+            f"{port}:{target_port}",
+        )
+
+    threads = [
+        threading.Thread(target=port_forward, args=["console", "http"]),
+        threading.Thread(target=port_forward, args=["balancerd", "pgwire"]),
+        threading.Thread(target=port_forward, args=["balancerd", "http"]),
+        threading.Thread(target=port_forward, args=["environmentd", "internal-sql"]),
+        threading.Thread(target=port_forward, args=["environmentd", "internal-http"]),
+    ]
+
+    for t in threads:
+        t.start()
+    while len(port_destinations) < len(threads):
+        pass
 
-        sleep(1)
-    assert node_port is not None
+    for target, target_port in port_destinations.items():
+        print(f"{target}: http://localhost:{target_port}/")
+    open(f"http://localhost:{port_destinations['console/http']}")
```

Comment:
Like right here. There is no guarantee the console http port has actually been forwarded yet.